### PR TITLE
doc: how to add a new parameter

### PR DESCRIPTION
**Pull request by jakmeier**
Add documentation on how new gas cost parameters should be added, as
reference for anyone implementing new features.

### CODE REVIEW

**Comment by jakmeier** [comment]
This is a piece of documentation we needed yesterday. Contributors have to figure this out on their own, which is not ideal. But instead of merging it in rushed, I think this one deserves a bit of a discussion.

First, is this type of step-by-step instructions a good fit for the architecture chapter? Or should go into misc?

Second, the principles I have written down are just from the top of my head and reflect my personal understanding. Would be great to get input from others.

cc @matklad  @akhi3030

**Comment by jakmeier** [comment]
Also, cc @blasrodri as a valued contributor who has gone through this process the hard way already. If you have inputs or spot an error in my step-by-step guide (I didn't test it step by step and haven't added a parameter in while) that would be most welcome! You are probably the best person to judge if this kind of docs are at all helpful.

File: docs/architecture/gas/parameter_definition.md [hunk 1]
```diff
@@ -6,8 +6,61 @@ IMPORTANT: This is not the final list of parameters, it contains the base values
...
 starting from version 53. To see all parameter values for a specific version at
-once, check out list of JSON snapshots generated in this directory:
+once, check out the list of JSON snapshots generated in this directory:
 [core/primitives/src/runtime/snapshots](https://github.com/near/nearcore/blob/master/core/primitives/src/runtime/snapshots).
 
-<!-- TODO: How to add a new parameter -->
 <!-- TODO: Runtime Config Store -->
+
+## How to Add a New Parameter
+
```
Comment (matklad) [inline docs/architecture/gas/parameter_definition.md:15]: Let's add "if you are feeling lost, open a thread at zulip/contract-runtime, we are here to help you"

**Comment by blasrodri** [comment]
> Also, cc @blasrodri as a valued contributor who has gone through this process the hard way already. If you have inputs or spot an error in my step-by-step guide (I didn't test it step by step and haven't added a parameter in while) that would be most welcome! You are probably the best person to judge if this kind of docs are at all helpful.

Will take a look and write my thoughts :)

File: docs/architecture/gas/parameter_definition.md [hunk 2]
```diff
@@ -6,8 +6,61 @@ IMPORTANT: This is not the final list of parameters, it contains the base values
...
+   gets included when compiling for nightly. After that, check `cargo check` and
+   `cargo test --no-run` with and without `features=nightly`.
+
+### What Gas Value Should the Parameter Have?
+For a first draft, the exact gas value used in the parameter is not crucial.
+Make sure the right set of parameters exists and try to set a number that roughly
+makes sense. This should be enough to enable discussions on the NEP around
+feasibility and usefulness of the proposed feature. If you are not sure, a good
+rule of thumb is 0.1 Tgas for each disk operation and at least 1 Tgas for each 
+ms of CPU time. Then round it up generously.
```
Comment (blasrodri) [inline docs/architecture/gas/parameter_definition.md:62]: do you want to add a reference to this as well? 
https://near.github.io/nearcore/architecture/gas/gas_profile.html

Comment (jakmeier) [reply docs/architecture/gas/parameter_definition.md:69]: Yeah that makes sense to also link to that. But probably better right after the list of principles?

File: docs/architecture/gas/parameter_definition.md (hunk 1 above)
Comment (jakmeier) [reply docs/architecture/gas/parameter_definition.md:15]: Good idea

File: docs/architecture/gas/parameter_definition.md [hunk 3]
```diff
@@ -6,8 +6,68 @@ IMPORTANT: This is not the final list of parameters, it contains the base values
...
+- Ideally, contract developers can easily understand what the cost is simply by
+  reading the name in a gas profile.
+
+The section on [Gas Profiles](./gas_profile.md#charging-gas) explains how to
+charge gas, please also take that into considerations when defining a new
+parameter.
+
+### Necessary Code Changes
+Adding the parameter in code involves several steps.
+1. Define the parameter by adding it to the list in `parameters.txt.`
```
Comment (blasrodri) [inline docs/architecture/gas/parameter_definition.md:38]:
```suggestion
1. Define the parameter by adding it to the list in [`parameters.txt`](https://github.com/near/nearcore/blob/master/core/primitives/res/runtime_configs/parameters.txt)
```

Comment (blasrodri) [reply docs/architecture/gas/parameter_definition.md:38]: I'm suggesting to either link it to the repo, or to do as in (2), where the path is also specified.

File: docs/architecture/gas/parameter_definition.md [hunk 4]
```diff
@@ -6,8 +6,68 @@ IMPORTANT: This is not the final list of parameters, it contains the base values
...
+   likely you have to update some testing code, such as
+   `ExtCostsConfig::test()`.
+6. To merge your changes into nearcore, you will have to hide your parameter
+   behind a feature flag. Add the feature to the `Cargo.toml` of each crate
+   touched in step 3 and 4 and hide the code behind `#[cfg(feature =
+   "protocol_feature_MY_NEW_FEATURE")]`. Do not hide code in step 2 so that
+   non-nightly builds can still read `parameters.txt`. Also add your feature as
+   a dependency on `nightly` in `core/primitives/Cargo.toml` to make sure it
+   gets included when compiling for nightly. After that, check `cargo check` and
+   `cargo test --no-run` with and without `features=nightly`.
```
Comment (blasrodri) [inline docs/architecture/gas/parameter_definition.md:61]: can this be converted into a make target? 
```
make check-parameters
```

**Review by matklad — APPROVED** [review]
Approving to unblock merging

File: docs/architecture/gas/parameter_definition.md (hunk 3 above)
Comment (jakmeier) [reply docs/architecture/gas/parameter_definition.md]: This file is already linked once on this page. So I added just the path.

**Comment by jakmeier** [comment]
I'll merge this now as we can always improve it in future PRs.

**Pull request merged**